### ISSUE TITLE
Send clipboard on demand only and/or send text

### ISSUE THREAD

**Issue by getzze** [GSConnect/gnome-shell-extension-gsconnect#1167]
The clipboard plugin is nice but it shares EVERY text with the phone. It would be nice to have the possibility to send the clipboard on demand, like this new feature request in KDE connect:
https://invent.kde.org/network/kdeconnect-kde/-/merge_requests/396

Otherwise (or in addition to the other feature), a button that opens a small text entry (like the sms window) and then sends this text to the phone would be great. On Android you can do something similar by selecting text and sharing it via KDE connect.

**Comment by andyholmes** [comment]
This should be easy already, there are settings in the preferences for automatic bi-directional sync:

![Screenshot from 2021-09-08 15-37-19](https://user-images.githubusercontent.com/1265208/132594837-c4927546-c63d-4ea5-9205-bf411316e989.png)

And there are two GActions, exposed over D-Bus, `device.clipboardPull` and `device.clipboardPush` that can be assigned to keyboard shortcuts. Were you thinking something else?

**Comment by getzze** [comment]
The GAction would have to be defined for a specific device, but I have several devices.
I was thinking of buttons in the menu:
![gsconnect](https://user-images.githubusercontent.com/6518000/132596651-3aa8e1ae-4d21-4ddc-803d-dc86e96d2fb2.png)

**Comment by andyholmes** [comment]
If you add `clipboardPush`/`clipboardPull` to the GSettings key:

    /org/gnome/shell/extensions/gsconnect/<device-id>/menu-actions

Whose default value is

    ["sms", "ring", "mount", "commands", "share", "photo", "keyboard"]

Do they show up in the menu then? You might have to restart the service for it to take effect (I can't remember).

**Comment by getzze** [comment]
I don't have the `menu-actions` key in `/org/gnome/shell/extensions/gsconnect/<device-id>/` and I don't know how to create it with dconf Editor.

**Comment by andyholmes** [comment]
Something like this should work:

    dconf write /org/gnome/shell/extensions/gsconnect/<device-id>/menu-actions '["sms", "ring", "mount", "commands", "share", "photo", "keyboard", "clipboardPull", "clipboardPush"]'

**Comment by getzze** [comment]
It's not working, the actions in the menu are unchanged (Messaging, Ring, ...)

In the meantime I remembered that you can send text with the Share action, it is so hidden that I had forgotten :)

**Comment by andyholmes** [comment]
Did you restart the service? I'm not sure there's any machinery for handling the menu actions changing on the fly.

    touch ~/.local/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/daemon.js

**Comment by getzze** [comment]
I logged out after sending the dconf command.
When I logged in, the previous items appeared in the menu of my device, then evething disappeared after 1s. Then some time later the items reappeared. But `clipboardPull` and `clipboardPush` never appeared.
I have another device connected that showed the normal items all the time.
 It seems it first showed the  items in cache, then tried to add the new ones but failed so all the menu disappeared and then it went back to default afterwards.

**Comment by andyholmes** [comment]
Yes, there's no real testing that case. The menu code could all probably use a good look.

**Comment by ferdnyc** [comment]
Huh. How strange, that it would end up out of order like that. Looking at the code, items are even inserted with an index number that corresponds to their position in the `menu-actions` list — unless they don't have an index (meaning it gets defaulted to `-1`, in which case they're _appended_ to the list. But I can't see why something would be inserted **forward** of its real position.

Hmm. Unless maybe the insertion somehow _wraps_ around — because, the index of `clipboardPush` would be 6, and at the time it's inserted the menu would only have 5 items (since `"commands"` appears to be missing).

...But, no, if that were the case then `"share"` and `"keyboard"` would have the same issue. (Oh, but wait, they're also out of order — or, rather, their order is swapped. How odd.)

**Comment by stdedos** [comment]
Did https://github.com/GSConnect/gnome-shell-extension-gsconnect/issues/1167#issuecomment-915649935 instructions change _recently_?

Only 2 (out of _a lot_) devices return results to `dconf dump /org/gnome/shell/extensions/gsconnect/device/ | grep -C5 menu` - and one of them (which might be one of the active ones) I cannot think how to "bind" its _folder-id_ to its _Encryption Info_ somehow.

(v66)

**Comment by andyholmes** [comment]
I doubt there have been any changes to the menu code, but since that setting was never really used, most devices won't have a value saved in dconf.

**Comment by stdedos** [comment]
Okay - I am using the panel (because the new gnome menu is ridiculous), and I don't see e.g. what the user https://github.com/GSConnect/gnome-shell-extension-gsconnect/issues/1167#issuecomment-2254462197 is seeing.

Are there any instructions on how to figure out which `/org/gnome/shell/extensions/gsconnect/device/<id>` is for an Connected device?
Or do I have to dig logs? 😕

**Comment by andyholmes** [comment]
There's no way in the UI to connect a device ID to a device, no. But I guess you could just set them all?

**Comment by stdedos** [comment]
> But I guess you could just set them all?

... I never really understood "that logic". Why can I not configure on each device that I install what I want to share or not and _if need be_ override per-connected-device, and INSTEAD I have to configure each device separately?

**Comment by andyholmes** [comment]
You can set up each device if you like, there's just no UI for configuring the menu.

**Comment by stdedos** [comment]
> You can set up each device if you like

What I am saying is "**I don't want to**". But the setup "each peer device in their own magical 'folder'" (without any 'medium-power-user' way to connect x-name-to-y-peer), and no "global config; override per peer device  (if needed/wanted)" really gives me no other option.
I have >20 "<id> folders" - despite having exactly 5 devices.

... anyhow - that's far deviating from _this issue's topic_. Apologies to all for my rant 🙏

**Comment by andyholmes** [comment]
You're welcome to submit a patch that changes [the default](https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/main/data/org.gnome.Shell.Extensions.GSConnect.gschema.xml#L65).

If you want to write a layered GSettings wrapper that allows for per-device and global fallbacks, that's definitely possible, but it is a decent amount of work. It would also need some design for the preferences window.

**Comment by zhelezov** [comment]
Maybe instead of encumbering the main menu, we can have a button in the Share dialog, like Send Clipboard next to Send URL? Could this be an easier way to implement this?
